### DISCUSSION TITLE
Automatically segments into sequences of maximum 145 for en_core_web_trf?

### DISCUSSION THREAD

**Discussion by seraphinatarrant** [explosion/spaCy#8546] [Help: Coding & Implementations] [answered]
I've been trying to use spacy to get contextual embeddings for individual tokens (which I assume is a standard use case since it is for the underlying huggingface). I notice that I can do this by accessing the  protected attribute via the following:
`doc = nlp(my_text)`
then
`doc._.trf_data.tensors[0]`

Where the 0 index seems to be the encoding with dimensions batch x seq_length x embed_dim and the 1 index seems to to be the output layer with dimensions batch x embed_dim.

But the confusing part is that it seems to automatically segment into max lengths of 145, but not in a way that is easy to merge?  So if the original doc is 162 wordpiece tokens, I end up with a tensor of (2, 145, 768) in which the first row is the first 145 tokens, but then the second one is not 17 tokens, as I would expect but is instead 52 tokens (followed by N pad tokens). So the first 35 tokens of the second row are repeated from the first row.

Is this standard behaviour and is the N tokens that overlap constant? Also, why is this, since BERT should take 512 (and even if not, why the overlap tokens)?

What are the recommendations for if I want to get the embeddings of token X if it is an arbitrary distance into the sequence? I can obviously just pre-segment into chunks of 145 but surely something is cleaner than this? 
I also feel it should be documented!

Thanks!

**Comment by adrianeboyd** [comment, marked as answer]
The transformer config in the default config and in the pretrained trf pipelines use overlapping strided spans, so that all tokens end up with left and right context where possible. Each span covers a certain number of spacy tokens, which map to some (larger) number of wordpiece tokens and each span overlaps by a certain number of spacy tokens with the next span. If the number of wordpiece tokens is longer than the max supported by the model, that span is truncated for the transformer processing steps.

The documentation for the `trf_data` data structure is here: https://spacy.io/api/transformer#transformerdata

And a few relevant threads:

* https://github.com/explosion/spaCy/issues/7032#issuecomment-777377338
* https://github.com/explosion/spaCy/discussions/6511

If you train your own transformer model, you can choose how to configure the spans under `[components.transformer.model.get_spans]`. If your docs are all shorter than the max length, you can use `doc_spans.v1` and then there is no overlap and the alignment would be a bit more straightforward.

**Comment by seraphinatarrant** [comment]
Thanks for the pointer to the relevant threads! I'd read the trf_data structure info but the striding information wasn't clear until you explained it, thanks so much. 
Looks like #6511 has what I want (and if not I can create it from this info)
